Upgrade OpenAI SDK to v1

tonybaloney opened the pull request:
## Summary of changes

- Open AI SDK v1 uses a client object, however it doesn't accept a deployment name as the parameter to the chat completions or embeddings methods, so we still need to create two clients. I've raised an issue with the Open AI team about this. In the meantime, it removed a lot of code to store state in global variables
- I've implemented both Azure OpenAI clients using a token provider instead of a runtime token, this means it will fetch a token for each session call and removed a lot of workaround code to refresh tokens
- The return types in the new SDK are Pydantic models. I've converted them back into dictionaries at the exit-point for the approach classes so it shouldn't leak too much of the abstraction
- The new Pydantic models have required fields that we didn't bother to mock/patch in the original test code. I wanted to make the mocked responses as realistic as possible (within reason), so I've included them and that meant updating all the snapshots
- I've changed the two mock fixtures for the chat completions and embeddings API to be fixture-factories that patch the method of the live client object instead of mocking the underlying OpenAI methods. This should make them more robust if OpenAI refactor their code

## TODO List

- [x] Update dependencies
- [x] Update prepdocs lib
- [x] Update backend service API
- [x] Update Chat Retrieve Read
- [x] Update Retrieve then Read
- [x] Test

Reviewers : please test as many bits of functionality as you can. I've tested this locally and deployed it to test in a production instance, but many hands find more bugs :-)

pamelafox commented:
Re tests- they did post some example mocks:
https://github.com/openai/openai-python/issues/715#issuecomment-1809203346
It might be useful to compare our approach to theirs. (I haven't had time yet myself)

tonybaloney commented:
> Re tests- they did post some example mocks: [openai/openai-python#715 (comment)](https://github.com/openai/openai-python/issues/715#issuecomment-1809203346) It might be useful to compare our approach to theirs. (I haven't had time yet myself)

thanks, I've done something very similar. Most of the remaining work is because the old SDK used to return dictionaries and the new one returns typed pydantic models

pamelafox reviewed (changes requested):
Krista says that we should use model to specify deployment, so you can use a single client, which I think is better for code complexity.

pamelafox commented on app/backend/approaches/chatreadretrieveread.py:
```diff
@@ -83,13 +83,33 @@ def __init__(
         self.query_speller = query_speller
         self.chatgpt_token_limit = get_token_limit(chatgpt_model)
 
+    @overload
```
I haven't used overload before, how do the additional signatures aid type checkers? Were these added due to a mypy error?

tonybaloney replied:
It was added because streaming=True returns a completely different type and instead of having Union's propagating from the return signature through the class and having big, messy type signatures this makes it simpler.

pamelafox commented on app/backend/approaches/chatreadretrieveread.py:
```diff
@@ -126,11 +146,9 @@ async def run_until_final_call(
-        chat_completion = await openai.ChatCompletion.acreate(
-            **chatgpt_args,
+        chat_completion: ChatCompletion = await self.openai_chat_client.chat.completions.create(
+            messages=messages,  # type: ignore
```
What doesn't it like about messages?

tonybaloney replied:
I've typed that. There's 1 remaining type: ignore that mypy is still complaining but I can't figure out why

tonybaloney replied:
See https://github.com/Azure-Samples/azure-search-openai-demo/pull/1017/commits/a9ef636438c0db11d2de659e5ebaacca74217316

mattmsft commented on app/backend/app.py:
```diff
@@ -250,20 +233,35 @@ async def setup_clients():
+            api_version="2023-07-01-preview",
+            azure_endpoint=f"https://{AZURE_OPENAI_SERVICE}.openai.azure.com",
+            azure_ad_token_provider=token_provider,
+            organization=OPENAI_ORGANIZATION,
```
do we need to set organization here?

tonybaloney replied:
No, we don't

tonybaloney commented on app/backend/approaches/chatreadretrieveread.py:
```diff
@@ -236,7 +257,8 @@ async def run_without_streaming(
+        chat_resp = chat_completion_response.model_dump()  # Convert to dict to make it JSON serializable
         chat_resp["choices"][0]["context"] = extra_info
         if overrides.get("suggest_followup_questions"):
             content, followup_questions = self.extract_followup_questions(chat_resp["choices"][0]["message"]["content"])
```
I looked at typing the chat completion chunks that this coroutine yields, but Choice doesn't have a context or session_state parameter.

pamelafox replied:
Ah that's because context and session_state are additions that we've defined in our Azure Chat App protocol. I don't know if it'd make sense to inherit from the openai Completion object to add those?

tonybaloney replied:
I'd have to define a bunch of custom models, which would remove the value of validating against OpenAI's API and make it kinda pointless.

tonybaloney commented:
@mattmsft @pamelafox comments have been addressed, please review again

pamelafox commented:
@tonybaloney I tested it and discovered an issue with follow-up questions. Previously, an event choice would either have a content string or have no content property at all, but now it can have "content": None. I've adjusted the code and test fixture to account for that.

pamelafox commented:
I tested it with non-Azure OpenAI with an existing environment and there's an issue due to this new logic in the approaches:

```
chatgpt_model = self.chatgpt_deployment if self.chatgpt_deployment else self.chatgpt_model
```

The problem is that AZURE_OPENAI_CHATGPT_DEPLOYMENT has a value even if you're using non-Azure OpenAI (defaults to "chat"). 

One way around that is to clear out the relevant variables in app.py, before they get passed into the approaches. This change worked for me:

```
    AZURE_OPENAI_CHATGPT_DEPLOYMENT = OPENAI_HOST == "azure" and os.getenv("AZURE_OPENAI_CHATGPT_DEPLOYMENT")
    AZURE_OPENAI_EMB_DEPLOYMENT = OPENAI_HOST == "azure" and os.getenv("AZURE_OPENAI_EMB_DEPLOYMENT")
```

The error can't be replicated in test, by the way, as it's an error from the server that the model "chat" cannot be found.

tonybaloney commented:
> @tonybaloney I tested it and discovered an issue with follow-up questions. Previously, an event choice would either have a content string or have no content property at all, but now it can have "content": None. I've adjusted the code and test fixture to account for that.

I added the `and content` on this line because I saw a bug where it would crash with follow up questions, but maybe I introduced another bug in the process? https://github.com/Azure-Samples/azure-search-openai-demo/pull/1017/commits/c2d20f22d8be8bcc44d13543f384c8dd7d1af043#diff-a346ba4129613c27f42a9a4de5439423af8ab6df6eb79eaf555204da3d588001R302 could this be simplified if we removed the `and content` in the if statement?

tonybaloney commented:
> I tested it with non-Azure OpenAI with an existing environment and there's an issue due to this new logic in the approaches:
> 
> ```
> chatgpt_model = self.chatgpt_deployment if self.chatgpt_deployment else self.chatgpt_model
> ```
> 
> The problem is that AZURE_OPENAI_CHATGPT_DEPLOYMENT has a value even if you're using non-Azure OpenAI (defaults to "chat").
> 
> One way around that is to clear out the relevant variables in app.py, before they get passed into the approaches. This change worked for me:
> 
> ```
>     AZURE_OPENAI_CHATGPT_DEPLOYMENT = OPENAI_HOST == "azure" and os.getenv("AZURE_OPENAI_CHATGPT_DEPLOYMENT")
>     AZURE_OPENAI_EMB_DEPLOYMENT = OPENAI_HOST == "azure" and os.getenv("AZURE_OPENAI_EMB_DEPLOYMENT")
> ```
> 
> The error can't be replicated in test, by the way, as it's an error from the server that the model "chat" cannot be found.

Good spot. Fixed

pamelafox commented on app/backend/requirements.in:
```diff
@@ -1,14 +1,15 @@
 quart
 quart-cors
-openai[datalib]==0.28.1
+openai[datalib]>=1.3.5,<2.0.0
```
Do we need the upper constraint?

tonybaloney replied:
the major version change indicates some sort of breaking change, so I'd rather have the constraint there just in case.

pamelafox replied:
We explicitly have a pip-compile step now, so I'm wondering why an upper constraint would be needed in the .in file, and why we'd have it for openai and not for other packages?

pamelafox commented on app/backend/app.py:
```diff
@@ -33,13 +32,13 @@
 CONFIG_AUTH_CLIENT = "auth_client"
 CONFIG_SEARCH_CLIENT = "search_client"
+CONFIG_OPENAI_CHAT_CLIENT = "openai_chat_client"
+CONFIG_OPENAI_EMBEDDINGS_CLIENT = "openai_embeddings_client"
```
Could you explain why you decided to keep two clients, even though they're the same object? I guess that's in case someone in the future wanted them to be different, it'd make it easier for them? It's not necessary now, right?

tonybaloney replied:
Fixed

pamelafox commented on app/backend/core/messagebuilder.py:
```diff
@@ -16,19 +23,30 @@ class MessageBuilder:
         """
-        self.messages.insert(index, {"role": role, "content": self.normalize_content(content)})
+        message: ChatCompletionMessageParam
+        if role == "user":
```
In theory this could be a dict with lookup based on role but I think it'd be about the same level as code and perhaps more confusing than what you have. I keep thinking that MessageBuilder is too complex and could be simplified but haven't wrapped my head around how.
So this works, just sharing my thoughts.

pamelafox commented on scripts/prepdocslib/embeddings.py:
```diff
@@ -139,18 +139,13 @@ def __init__(
+    async def create_client(self) -> AsyncOpenAI:
+        return AsyncAzureOpenAI(
+            azure_endpoint=f"https://{self.open_ai_service}.openai.azure.com",
+            azure_deployment=self.open_ai_deployment,
```
Oh, interesting, you set deployment at the client level. I find that slightly unintuitive since I'm accustomed to deployment at function level but it does admittedly work well for this particular class, in terms of reducing the Azure/non-Azure differences, so this works. 👍🏼

pamelafox commented on scripts/requirements.in:
```diff
@@ -1,9 +1,10 @@
 azure-storage-blob
 azure-storage-file-datalake
-openai[datalib]==0.28.1
+openai[datalib]>=1.3.5,<2.0.0
```
You can remove upper here too

pamelafox replied:
Also you can now remove the ignore from dependabot of openai package, as I was only ignoring it while we weren't on 1.0

pamelafox commented on tests/test_app.py:
```diff
@@ -3,13 +3,34 @@
 
 
+def fake_response(http_code):
+    return Response(http_code, request=Request(method="get", url="https://foo.bar/"))
```
I'm confused by foo.bar, how is that connected to the tests?

tonybaloney replied:
It isn't, but it is a required property of the Request option so I had to put something there

pamelafox commented on tests/test_chatapproach.py:
```diff
@@ -1,37 +1,46 @@
-    chat_approach = ChatReadRetrieveReadApproach(
-        None, "", "gpt-35-turbo", "gpt-35-turbo", "", "", "", "", "en-us", "lexicon"
+@pytest.fixture
+def chat_approach():
```
Oh great, you made it a fixture! I was thinking this should be a fixture, given its constant use. Thanks!

pamelafox commented on tests/test_prepdocs.py:
```diff
@@ -84,12 +107,23 @@ async def mock_create(*args, **kwargs):
 
 
+def fake_response(http_code):
+    return Response(http_code, request=Request(method="get", url="https://foo.bar/"))
```
The foo.bar here also seems like placeholder, maybe make this look like an openai URL?

pamelafox reviewed (commented):
Done with my deploy tests and code review. No big items, just minor comments.

mattmsft reviewed (commented):
LGTM
